### PR TITLE
Contributors guide

### PR DESCRIPTION
**Pull request by clyne**
This PR partially (mostly) addresses #19. It provides a general, but detailed, guide to contributing to Project Pythia, and provides links to repo-specific contributors information (i.e. how to contribute to Foundations). The general guide covers the different ways to contribute (including how to submit an external link); setting up git, github, and conda; and how to use the "forking workflow"

N.B. changes to the foundations contributor's guide are forthcoming

### CODE REVIEW

**Comment by clyne** [comment]
@andersy005, any guidance on how to trouble shoot these lint errors? Can I run the linter locally...Thanks!

**Comment by andersy005** [comment]
> @andersy005, any guidance on how to trouble shoot these lint errors? Can I run the linter locally...Thanks!


@clyne, Try the following

- Add `pre-commit` to the list of dependencies in `ci/environment.yml` and run `conda env update -f ci/environment.yml`
- Install `pre-commit` with `conda activate pythia && pre-commit install`
- Run `pre-commit run --all-files` to fix files from previous commits
- The `pre-commit` hooks should run automatically for any subsequent commits.

**Comment by clyne** [comment]
Thanks @andersy005 ! That worked. Is there a reason we don't have this baked in to our build environment? We seem to for the foundations portal...

**Comment by andersy005** [comment]
> Thanks @andersy005 ! That worked. Is there a reason we don't have this baked in to our build environment? We seem to for the foundations portal...

You are welcome! I think we just forgot to include it in our local build environment

File: content/pages/contributing.md
Comment (andersy005) [inline content/pages/contributing.md:190]:
```suggestion
conda update conda
```

**Comment by clyne** [comment]
How about we add it, and I'll update the contributor's guide while I'm working on it?

**Comment by andersy005** [comment]
> How about we add it, and I'll update the contributor's guide while I'm working on it?

👍🏽 Sounds good

**Review by kmpaul — APPROVED** [review]
Looks good.  Thanks, @clyne!

File: content/pages/contributing.md
Comment (jukent) [inline content/pages/contributing.md:47]: These steps are different/there are 2 paths now that #87 is merged. The user can also:

1. Navigate to the External Links Gallery.
2. Click the "Submit a Link" button, which will redirect them to a new PR template.
3. Fill out that template with the information relevant to their Python learning resource.

**Review by jukent — CHANGES_REQUESTED** [review]

File: content/pages/contributing.md
Comment (clyne) [reply content/pages/contributing.md:47]: Nice! I updated the documentation per above.

**Review by jukent — APPROVED** [review]

File: content/pages/contributing.md
Comment (jukent) [reply content/pages/contributing.md:47]: Great!

**Review by andersy005 — DISMISSED** [review]

File: content/pages/contributing.md
Comment (brian-rose) [inline content/pages/contributing.md:189]: We already have detailed instructions for installing conda and setting up a Python environment in the Foundations book [here](https://foundations.projectpythia.org/foundations/conda.html)

It would be good link to this.

File: content/pages/contributing.md
Comment (brian-rose) [inline content/pages/contributing.md:266]: Replace GeoCAT with `PYTHIA_REPO_NAME`

File: content/pages/contributing.md
Comment (brian-rose) [inline content/pages/contributing.md:357]: Do we want to discuss the use of the precommit hooks here?

File: content/pages/contributing.md
Comment (brian-rose) [inline content/pages/contributing.md:392]: Replace GeoCAT

File: content/pages/contributing.md
Comment (brian-rose) [inline content/pages/contributing.md:400]: remove extra ` after `pythia-foundations`

File: content/pages/contributing.md
Comment (brian-rose) [inline content/pages/contributing.md:432]: Congradulations ---> Congratulations

**Review by brian-rose — CHANGES_REQUESTED** [review]

File: content/pages/contributing.md
Comment (clyne) [reply content/pages/contributing.md:189]: Good catch. Thanks!

File: content/pages/contributing.md
Comment (clyne) [reply content/pages/contributing.md:266]: Oops. Fixed.

File: content/pages/contributing.md
Comment (clyne) [reply content/pages/contributing.md:392]: Got it.

File: content/pages/contributing.md
Comment (clyne) [reply content/pages/contributing.md:400]: done

File: content/pages/contributing.md
Comment (clyne) [reply content/pages/contributing.md:357]: What information do you think should be added?

**Comment by brian-rose** [comment]
My overall comment is that I really like this verbose, newbie-friendly guide. However I think that *some* of this material should migrate over to the Foundations book at some point, where we hope to have general tutorials on things like use of git, opening PRs, etc. 

Once those tutorials are in place, this guide could be shortened and cross-linked to the relevant Foundations pages.

The only difference would be that the Foundations tutorials would be aimed at more general contributions, not just contributing to Pythia itself.

**Comment by clyne** [comment]
> My overall comment is that I really like this verbose, newbie-friendly guide. However I think that _some_ of this material should migrate over to the Foundations book at some point, where we hope to have general tutorials on things like use of git, opening PRs, etc.
> 
> Once those tutorials are in place, this guide could be shortened and cross-linked to the relevant Foundations pages.
> 
> The only difference would be that the Foundations tutorials would be aimed at more general contributions, not just contributing to Pythia itself.

I completely agree that some of this content should be moved to foundations and referenced here. I'll open a new issue.

File: content/pages/contributing.md
Comment (clyne) [reply content/pages/contributing.md:360]: I've added a few sentences about pre-commit hooks. CC'ing @andersy005 and asking to re-review (this topic)

File: content/pages/contributing.md
Comment (brian-rose) [inline content/pages/contributing.md:369]: This should specify that you need to do
``` 
git add PATH_TO_NEW_FILE
```
one more time to add the new changes to your commit, and only then do another `git commit`.

File: content/pages/contributing.md
Comment (brian-rose) [reply content/pages/contributing.md:360]: See my comment below, I think we need to spell out the details: git add, git commit, let the hooks make changes, then git add, git commit again.

**Review by brian-rose — CHANGES_REQUESTED** [review]

File: content/pages/contributing.md
Comment (clyne) [reply content/pages/contributing.md:369]: Hmm. I didn't think 'git add' was needed unless you added a new file to the repo. I've only had to run 'git commit'. Is there a possibility that the hook will add a new file, or is my understanding incorrect?

Comment (brian-rose) [reply content/pages/contributing.md:369]: In my experience, the hook will make changes, but you still need to `git add` those changes for them to be included in the commit.

It may be that there are some configuration issues with `precommit` that change this behavior. @andersy005 ?

**Comment by clyne** [comment]
Re "git add": I can certainly add info on what to do if you've explicitly created a new file. That makes sense. I just want to make sure I understand the need for 'git add' in the content of our hooks. Thanks

File: content/pages/contributing.md
Comment (brian-rose) [reply content/pages/contributing.md:369]: It's not creating a new file, it's making modifications to an existing file. git will then treat these modifications just like any other. Until you explicitly `add` them to the staging area, those changes are not part of the commit.

Comment (clyne) [reply content/pages/contributing.md:369]: Uhh. Not trying to be difficult, but this doesn't jive with my understanding (and use of git). My understanding/usage:

1. Use 'git add' the first time you add a file to your local .git repo. This step need never be repeated again (I think)
2. Thereafter use 'git commit' whenever changes are made to a file that you want to update the local repo with.

Do you use different (preferred) process? Just trying to understand

Comment (brian-rose) [reply content/pages/contributing.md:369]: Ok, I think I see the disconnect.

Any git commit is a two-step process. Files are added to the "staging area" with `git add FILENAME`, and anything in the staging area is committed with `git commit`.

It's possible to fast-forward through the two-step process by doing
```
git commit FILENAME
```
which is equivalent to 
```
git add FILENAME
git commit
```

But if you make some changes to a previously committed file, and then just do 
```
git commit
```
(without specifying which files to commit), then nothing will happen.

Comment (brian-rose) [reply content/pages/contributing.md:369]: So in the instructions where you say "you may need to run `git commit` a second time",  maybe better to say "run `git commit FILENAME` a second time", or spell out the two-stage process.

Does that make more sense?

Comment (brian-rose) [reply content/pages/contributing.md:369]: Usually @dopplershift is better at explaining this kind of stuff...

Comment (andersy005) [reply content/pages/contributing.md:369]: I personally use both commands `git add` & `git commit` regardless of whether it's a new file or an existing one... :)

Comment (clyne) [reply content/pages/contributing.md]: Now I see the disconnect too. Basically I forgot the file name.  Please see latest revisions.

Comment (clyne) [reply content/pages/contributing.md]: > So in the instructions where you say "you may need to run `git commit` a second time", maybe better to say "run `git commit FILENAME` a second time", or spell out the two-stage process.
> 
> Does that make more sense?

It does make more sense. I think that is what I have in there. Maybe we crossed wires. Could you please refresh and look?

Comment (clyne) [reply content/pages/contributing.md]: > I personally use both commands `git add` & `git commit` regardless of whether it's a new file or an existing one... :)

Is there a preference for one over the other, or is it fine to leave it as is?

Comment (brian-rose) [reply content/pages/contributing.md]: Ok, looks good to me now! I think the instructions are clear and the new callout note helps make things clear. 

This has been interesting. I guess I learned one workflow year ago (git add, git commit), and I've never revisited it until now.

**Review by brian-rose — APPROVED** [review]

**Comment by clyne** [comment]
Just need some of you folks who gave thumbs up, looks good, etc. to actually approve :-). Tagging @andersy005 @kmpaul @jukent 

Nag, nag, nag..

**Review by jukent — APPROVED** [review]

**Comment by brian-rose** [comment]
Looks like this needs approval specifically from @kmpaul and/or @ktyle as code owners.

**Review by ktyle — APPROVED** [review]

**Review by brian-rose — APPROVED** [review]

**Comment by brian-rose** [comment]
Ok the merge button is still red even with 3 explicit approvals, because of our relic code-owners setup that will be fixed with #105.

Either @kmpaul can approve and merge this, or if he's busy elsewhere I'll just force-merge later today.

**Comment by clyne** [comment]
This PR seems stuck in the "Read the Docs" build. Anyone (@andersy005 , @kmpaul ) no how to unstick it?

**Comment by jukent** [comment]
Should we add joining the email list to the contributor's guide?

If you would like join the projectpythia@ucar.edu Google group to be informed of updates and events in the seminar series, please do so by following this [link](https://groups.google.com/a/ucar.edu/g/projectpythia/about) and clicking "Join Group" next to the group name.

**Comment by jukent** [comment]
> This PR seems stuck in the "Read the Docs" build. Anyone (@andersy005 , @kmpaul ) no how to unstick it?

When I click "details" it says the build completed. Not sure why it is hanging.

**Comment by brian-rose** [comment]
Yep, same over at #105. I tried closing and reopening to trigger another build on readthedocs, but still hangs forever. Annoying but seems clear that it has nothing to do with the content of these PRs.

**Comment by brian-rose** [comment]
I think I got this working again by going over to Read the Docs and clicking on "Resync Webhooks" under the admin tab.

**Comment by brian-rose** [comment]
Now if we merge main in one more time (which now includes the changes to codeowners from #105), I guess that the review requirements will be met automatically. 

Let me try that.

**Review by andersy005 — APPROVED** [review]

**Comment by brian-rose** [comment]
> Should we add joining the email list to the contributor's guide?
> 
> If you would like join the [projectpythia@ucar.edu](mailto:projectpythia@ucar.edu) Google group to be informed of updates and events in the seminar series, please do so by following this [link](https://groups.google.com/a/ucar.edu/g/projectpythia/about) and clicking "Join Group" next to the group name.

@jukent given the multiple rounds of review on this particular PR, I'd prefer to merge this as-is now. Can you please open a new issue / PR to add the email list info?

**Comment by clyne** [comment]
I got the green lights, and I squashed and merged before anything else could happen!!! Thanks for sorting this out, @brian-rose

**Pull request merged**
